src/static/helperTemplates/slas.ts.hbs

```diff
@@ -0,0 +1,272 @@
+ * @param codeVerifier - random string to use as a code verifier
+ * @returns code challenge
+ */
+export const generateCodeChallenge = async (
+  codeVerifier: string
+): Promise<string> => {
+  const urlSafe = (input: string) =>
+    input.replace(/\+/g, "-").replace(/\//g, "_").replace(/=/g, "");
+
+  const crypto = await import("crypto");
+  const challenge = urlSafe(
+    crypto.default.createHash("sha256").update(codeVerifier).digest("base64")
+  );
+
+  if (challenge.length === 0) {
+    throw new Error("Problem generating code challenge");
+  }
+
+  return challenge;
+};
+
+/**
+ * Wrapper for the authorization endpoint. For federated login (3rd party IDP non-guest), the caller should redirect the user to the url in the url field of the returned object. The url will be the login page for the 3rd party IDP and the user will be sent to the redirectURI on success. Guest sessions return the code and usid directly with no need to redirect.
+ * @param slasClient - a configured instance of the ShopperLogin SDK client
+ * @param codeVerifier - random string created by client app to use as a secret in the request
+ * @param parameters - Request parameters used by the `authorizeCustomer` endpoint.
+ * @param parameters.redirectURI - the location the client will be returned to after successful login with 3rd party IDP. Must be registered in SLAS.
+ * @param parameters.hint? - optional string to hint at a particular IDP. Guest sessions are created by setting this to 'guest'
+ * @param parameters.usid? - optional saved SLAS user id to link the new session to a previous session
+ * @returns login url, user id and authorization code if available
+ */
+export async function authorize(
+  slasClient: ShopperLogin,
+  codeVerifier: string,
+  parameters: {
+    redirectURI: string;
+    hint?: string;
+    usid?: string;
+  }
+): Promise<{ code: string; url: string; usid: string }> {
+  const codeChallenge = await generateCodeChallenge(codeVerifier);
+
+  // Create a copy in order to modify behavior of the fetch call
+  // We do not want to redirect to redirectURI so manually control redirect
+  const slasClientCopy = new ShopperLogin(slasClient.clientConfig);
+  slasClientCopy.clientConfig.fetchOptions = {
+    ...slasClient.clientConfig.fetchOptions,
+    redirect: "manual",
+  };
+
+  const options = {
+    parameters: {
+      client_id: slasClient.clientConfig.parameters.clientId,
+      code_challenge: codeChallenge,
+      ...(parameters.hint && { hint: parameters.hint }),
+      organizationId: slasClient.clientConfig.parameters.organizationId,
+      redirect_uri: parameters.redirectURI,
+      response_type: "code",
+      ...(parameters.usid && { usid: parameters.usid }),
+    },
+  };
+
+  const response = await slasClientCopy.authorizeCustomer(options, true);
+  const redirectUrl = response.headers?.get("location") || response.url;
+
+  if (!redirectUrl) {
+    throw new Error("Authorization failed");
```

Comment:
Hey @joeluong-sfcc – on `commerce-sdk-isomorphic`, one thing thats blocking us from adopting the helpers is lack of access to the underlying errors.

This first call may fail for a variety of reasons:

1. SLAS miss configured (bad tenant, bad client, bad redirect URI)
2. Bad code challenge (repeated, invalid)
3. No network, transient network error etc.

I believe the SDK already has a pattern of attaching the `response` object to the error so that callers can handle it and display an appropriate error message to the user.

Can we use it here as well? <3 <3 <3

Reply:
`ResponseError` [exists in the core package](https://github.com/SalesforceCommerceCloud/commerce-sdk-core/blob/096f14d1300b3a26cce2ce8abd62ac5080d65035/src/base/staticClient.ts#L51), and can be used here.